### PR TITLE
feat: INTMDB-1023 private endpoint

### PR DESCRIPTION
**Pull request by Adelmar92**
## Proposed changes
We have made significant improvements to our Private Endpoint resource structure to enhance its usability and maintainability.

Previously, the MongoDB::Atlas::PrivateEndpoint resource was responsible for handling three distinct tasks:

Creating and deleting an Atlas Private Endpoint service.
Adding and removing an Atlas Private Endpoint to/from a Private Endpoint Service.
Creating and deleting an AWS private endpoint.
This approach had several drawbacks:

The complexity of having a single resource manage three different functions made it challenging to maintain.
Users found it limiting to manage an AWS Private Endpoint within an Atlas resource, which restricted certain configurations.
When a resource was created, there was no visibility into the progress of each component.
To address these issues, we have introduced a new resource, MongoDB::Atlas::PrivateEndpointService. This resource is specifically designed to create an Atlas Private Endpoint Service without any preconfiguration, allowing for subsequent Private Endpoint configuration with AWS or any other provider.

Additionally, we have added the MongoDB::Atlas::PrivateEndpointAWS resource to focus solely on adding a Private Endpoint to the Service.

	- MongoDB::Atlas::PrivateEndpointService: new, resource responsible for creating a Private Endpoint Service, unconfigured, and then using it to configure a Private Endpoint with AWS or any other provider
	- MongoDB::Atlas::PrivateEndpointAWS: the current resource is modified, it is responsible for adding a privateEdnpoint to the Service

we added a new feature so we can give the user the option to use both resources at the same time, while they are migrating to the new structure. also this was due to a bug present on Cloud formation where the users are not able to revert to an older resource version. 

As a result, users will now configure a Private Endpoint by defining these resources separately:

Stack:

Atlas Private Endpoint Service
AWS Private Endpoint
Atlas Private Endpoint

_Jira ticket: [INTMDB-1023](https://jira.mongodb.org/browse/INTMDB-1023)

In this test ive created the three resources in a single stac to complete a private endpoint configuration 
image

image

image

Link to any related issue(s): 

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as
  expected)
- [ ] This change requires a documentation update

## Manual QA performed:

- [x] cfn invoke for each of CRUDL/cfn test
- [x] Updated resource in  [example](https://github.com/mongodb/mongodbatlas-cloudformation-resources/tree/master/examples)
- [x] Published to AWS private registry
- [x] Used the template in [example](https://github.com/mongodb/mongodbatlas-cloudformation-resources/tree/master/examples) to create and update a stack in AWS
- [x] Deleted stack to ensure resources are deleted
- [x] Created multiple resources in same stack
- [x] Validated in Atlas UI
- [x] Included screenshots

## Required Checklist:

- [ ] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added any necessary documentation (if appropriate)
- [ ] I have run `make fmt` and formatted my code
- [ ] For CFN Resources: I have released by changes in the private registry and proved by change
  works in Atlas

### CODE REVIEW

**Comment by maastha** [comment]
Hi @Adelmar92 could you please resolve all conflicts and let me know when this is ready for review? 
Also please raise smaller PRs in the future as they are much more efficient to review:)

If the changes are related, one approach we sometimes use is to have separate PRs for each change but with related base branches in the PR.

**Comment by Adelmar92** [comment]
> Hi @Adelmar92 could you please resolve all conflicts and let me know when this is ready for review? Also please raise smaller PRs in the future as they are much more efficient to review:)
> 
> If the changes are related, one approach we sometimes use is to have separate PRs for each change but with related base branches in the PR.

Hi @maastha i know, about the 2 resources in a single pr, sorry about that, both of the resources should be used togheder, and the developing process was extremely tedious having them separated. i had to unify them in a single branch to make things simplier.

File: cfn-resources/private-endpoint-service/cmd/resource/resource.go
Comment (AgustinBettati) [inline cfn-resources/private-endpoint-service/cmd/resource/resource.go:76]: It there a reason we limit the cloud provider to AWS in private endpoint service, but leave it configurable in private endpoint?

File: cfn-resources/private-endpoint-service/cmd/resource/resource.go
Comment (AgustinBettati) [inline cfn-resources/private-endpoint-service/cmd/resource/resource.go:32]: nit: we can remove this alias and simply use progressevent

File: cfn-resources/private-endpoint-service/cmd/resource/resource.go
Comment (AgustinBettati) [inline cfn-resources/private-endpoint-service/cmd/resource/resource.go:34]: is there a reason we are not making use of the new go SDK in the private-endpoint-service?

**Review by AgustinBettati — CHANGES_REQUESTED** [review]
Left an initial set of comments, given the size of the PR I imagine we might need some iterations

**Comment by AgustinBettati** [comment]
Additionally, please include screenshots for the required checklist point ```For CFN Resources: I have released my changes in the private registry and proved by change works in Atlas```

**Comment by AgustinBettati** [comment]
We must also take into account that since these are breaking changes to the `MongoDB::Atlas::PrivateEndpoint` a major release will be needed.

File: cfn-resources/private-endpoint-service/cmd/resource/resource.go
Comment (Adelmar92) [reply cfn-resources/private-endpoint-service/cmd/resource/resource.go:72]: so in mi mind the user might need a service for other provider,
to create a complete private endpoint you need:
1- create the service for an specific provider (private-endpoint-service resource)
2-create the aws private endpoint (this the user must use the aws private endpoint cfn resource)
3-create an atlas private endpoint

leaving the CloudProvider we can leave a door so the user can at least create the step 1 for the other prividers, and then setup tha azure or any other connection by hand

File: cfn-resources/private-endpoint-service/cmd/resource/resource.go
Comment (Adelmar92) [reply cfn-resources/private-endpoint-service/cmd/resource/resource.go:32]: it was just for timing issues, this resource already existed (this is the former private-endpoint) so since there was a task for this change, i just removed the private endpoint, and the aws private endpoint from this one

File: cfn-resources/private-endpoint-service/cmd/resource/resource.go
Comment (AgustinBettati) [reply cfn-resources/private-endpoint-service/cmd/resource/resource.go:72]: From what you mention in point 1, we would leave the provider configurable in private-endpoint-service resource. However, as of the current implementation private-endpoint-service is fixed to AWS, while private-endpoint has a configurable provider property so I'm confused.

File: .github/workflows/contract-testing.yaml
Comment (andreaangiolillo) [inline .github/workflows/contract-testing.yaml:65]: was this split asked by a customer? Why did we decide to split the resource?

File: cfn-resources/private-endpoint-service/cmd/resource/resource.go
Comment (AgustinBettati) [inline cfn-resources/private-endpoint-service/cmd/resource/resource.go:249]: particular reason for a copy vs a normal assign statement?
On another hand, name of `completeByConnection` is hard to understand, would favor `populatePrivateEndpointProperties`

File: cfn-resources/private-endpoint-service/test/inputs_1_create.template.json
Comment (AgustinBettati) [inline cfn-resources/private-endpoint-service/test/inputs_1_create.template.json:4]: Is there a reason contract test has very few properties? We should add all possible fields to avoid regressions in the future.

**Review by AgustinBettati — COMMENTED** [review]
some smaller details

File: cfn-resources/private-endpoint-service/cmd/resource/resource.go
Comment (Adelmar92) [reply cfn-resources/private-endpoint-service/cmd/resource/resource.go:72]: yes you are right on this one, they are switch... ill change that

File: cfn-resources/private-endpoint-service/cmd/resource/resource.go
Comment (Adelmar92) [reply cfn-resources/private-endpoint-service/cmd/resource/resource.go:249]: i remember i had some issues with the pointers, they were replaced in the test, let me check again and ill let you know

File: .github/workflows/contract-testing.yaml
Comment (Adelmar92) [reply .github/workflows/contract-testing.yaml:65]: Originally this resource was ment to be be split, private endpoint, and private endpoint service were two separated resources (since in terraform and on atlas they are devided), also this resource controlls an AWS private endpoint, that is quite comples, so now is way easier to maintain because we dont have to worry about the AWS private endpoint

Comment (andreaangiolillo) [reply .github/workflows/contract-testing.yaml:65]: I am still not sure why we are making this change now. Customers using the resource will be confused by this change. Was this requested by @Zuhairahmed? What happens to customers currently using the resource and upgrading to the latest ones? 

We already have many resources (more than 30) and adding a new one should be done carefully.

**Review by andreaangiolillo — DISMISSED** [review]
waiting for Zuhair to address my comment

File: cfn-resources/private-endpoint-service/cmd/resource/resource.go
Comment (andreaangiolillo) [inline cfn-resources/private-endpoint-service/cmd/resource/resource.go:32]: This resource should use the autogenerated SDK instead

File: cfn-resources/private-endpoint/cmd/resource/resource.go
Comment (AgustinBettati) [inline cfn-resources/private-endpoint/cmd/resource/resource.go:147]: Is there a reason why the private endpoint Model has an Id property that will be the same as InterfaceEndpointId?

File: cfn-resources/private-endpoint-service/cmd/resource/resource.go
Comment (AgustinBettati) [inline cfn-resources/private-endpoint-service/cmd/resource/resource.go:305]: Make sure to cleanup println in favor of logging when appropriate

**Review by AgustinBettati — COMMENTED** [review]
additional comments from latest changes

File: cfn-resources/private-endpoint/mongodb-atlas-privateendpoint.json
Comment (AgustinBettati) [inline cfn-resources/private-endpoint/mongodb-atlas-privateendpoint.json:3]:
```suggestion
    "description": "DEPRECATED- USE MongoDB::Atlas::PrivateEndpointService
    and MongoDB::Atlas::PrivateEndpointAWS instead, The Private Endpoint creation flow consists of the creation of three related resources in the next order: 1. Atlas Private Endpoint Service 2. Aws VPC private Endpoint 3. Atlas Private Endpoint",
```

File: cfn-resources/private-endpoint-aws/mongodb-atlas-privateendpointaws.json
Comment (AgustinBettati) [inline cfn-resources/private-endpoint-aws/mongodb-atlas-privateendpointaws.json:5]:
```suggestion
    "sourceUrl": "https://github.com/mongodb/mongodbatlas-cloudformation-resources/tree/master/cfn-resources/private-endpoint-aws",
    "documentationUrl": "https://github.com/mongodb/mongodbatlas-cloudformation-resources/blob/master/cfn-resources/private-endpoint-aws/README.md",
```

File: cfn-resources/cfn-testing-helper.sh
Comment (AgustinBettati) [inline cfn-resources/cfn-testing-helper.sh:104]: why is this being removed?

File: cfn-resources/private-endpoint-aws/CHANGELOG.md
Comment (AgustinBettati) [inline cfn-resources/private-endpoint-aws/CHANGELOG.md:21]:
```suggestion
Additionally, we have defined a new MongoDB::Atlas::PrivateEndpointAWS resource to focus solely on adding a Private Endpoint to the Service.
```
This was outdated right?

File: cfn-resources/private-endpoint-aws/CHANGELOG.md
Comment (AgustinBettati) [inline cfn-resources/private-endpoint-aws/CHANGELOG.md:24]:
```suggestion
	- MongoDB::Atlas::PrivateEndpointAWS: the current resource is modified, it is responsible for adding a privateEdnpoint to the Service
```

File: cfn-resources/private-endpoint-aws/CHANGELOG.md
Comment (AgustinBettati) [inline cfn-resources/private-endpoint-aws/CHANGELOG.md:32]: this would also need adjustment, we can also mention that the new MongoDB::Atlas::PrivateEndpointAWS is specific to AWS

File: cfn-resources/private-endpoint-aws/CHANGELOG.md
Comment (AgustinBettati) [inline cfn-resources/private-endpoint-aws/CHANGELOG.md:1]: I believe this would be more relevant if placed in existingprivate endpoint resource, as users that are making use of that resource will likely want to know why it was deprecated.

File: cfn-resources/private-endpoint/docs/README.md
Comment (AgustinBettati) [inline cfn-resources/private-endpoint/docs/README.md:45]: does not seems intentional, we have to adjust the json schema here for the fix

File: cfn-resources/private-endpoint-aws/upgradeguidev2/V2-UpgradeGuide.md
Comment (AgustinBettati) [inline cfn-resources/private-endpoint-aws/upgradeguidev2/V2-UpgradeGuide.md:1]: As mentioned in other comment, given this is relevant for users using existing PrivateEndpoint resource, would place migration guide there.

File: examples/private-endpoint/privateEndpoint.json
Comment (AgustinBettati) [inline examples/private-endpoint/privateEndpoint.json:32]: should we leave the private endpoint example as it was, and create a new one using new resources? will also be helpful for users to compare.

**Review by AgustinBettati — COMMENTED** [review]
details worth addressing before we merge

File: cfn-resources/private-endpoint/CHANGELOG.md
Comment (AgustinBettati) [inline cfn-resources/private-endpoint/CHANGELOG.md:24]: We should mention that old resource is being deprecated here, and mention PrivateEndpointAWS as new.
Another doubt I have is that we are referring to the new structure as V2, but I believe all of the resource including old one (private-endpoint, private-endpoint-aws, private-endpoint-service) will be published with 2.0.0. So maybe it is worth rephrasing from "V2 Upgrade Guide" "New Structure Upgrade Guide", same in examples.

**Review by AgustinBettati — COMMENTED** [review]
documentation comment, we should be good to merge after this.

File: .github/workflows/contract-testing.yaml
Comment (Zuhairahmed) [reply .github/workflows/contract-testing.yaml:65]: yes @andreaangiolillo we synced on this to better align with Terraform/other tools functionality. @Adelmar92 to be clear we are keeping existing Private Endpoint resource as is and just adding new AWS Private Endpoint Resource and new Private Endpoint Service Resource right?

Comment (Zuhairahmed) [reply .github/workflows/contract-testing.yaml:65]: looking at below this appears to be the case, there should be no immediate change to quick start templates or breaking changes with this. also agree we will need new L1 constructors to support these 2 new resources

**Review by AgustinBettati — APPROVED** [review]
LGTM, thanks for addressing comments during several iterations.

**Pull request merged**
